File: leetcode/algorithms/501_find_mode_in_binary_search_tree/main.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class FindModeInBinarySearchTree:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        map = {}
        result = []

        def checkNode(node: TreeNode):
            if node == None:
                return

            if node.left != None:
                checkNode(node.left)

            if node.right != None:
                checkNode(node.right)

            if map.get(node.val) == None:
                map.update({node.val: 1})
            else:
                map.update({node.val: map.get(node.val) + 1})

        checkNode(root)

        for i in range(len(list(map.values()))):
            if list(map.values())[i] >= max(list(map.values())):
                result.append(list(map.keys())[i])

        return result


    # Solution
    prev = None
    max_count = 0
    current_count = 0
    result = []

    def solution(self, root: Optional[TreeNode]) -> List[int]:
        self.dfs(root)
        return self.result

    def dfs(self, node):
        if not node:
            return

        self.dfs(node.left)

        self.current_count = 1 if node.val != self.prev else self.current_count + 1

        if self.current_count == self.max_count:
            self.result.append(node.val)
        elif self.current_count > self.max_count:
            self.result = [node.val]
            self.max_count = self.current_count

        self.prev = node.val

        self.dfs(node.right)

File: leetcode/algorithms/501_find_mode_in_binary_search_tree/main.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class FindModeInBinarySearchTree {
    /**
     * Definition for a binary tree node.
     */
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int[] findMode(TreeNode root) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        List<Integer> result = new ArrayList<Integer>();

        checkNode(map, root);

        System.out.println();
        int max = Collections.max(map.values());

        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() >= max) {
                result.add(entry.getKey());
            }
        }

        return result.stream().mapToInt(v -> v).toArray();
    }

    public void checkNode(Map<Integer, Integer> map, TreeNode node) {
        if (node == null) {
            return;
        }

        if (node.left != null) {
            checkNode(map, node.left);
        }

        if (node.right != null) {
            checkNode(map, node.right);
        }

        if (map.get(node.val) == null) {
            map.put(node.val, 1);
        } else {
            map.put(node.val, map.get(node.val) + 1);
        }
    }


    // Solution
    // Solution 1: In-order traversal
    private int currVal;
    private int currCount = 0;
    private int maxCount = 0;
    private int modeCount = 0;

    private int[] modes;

    private void handleValue(int val) {
        if (val != currVal) {
            currVal = val;
            currCount = 0;
        }
        currCount++;
        if (currCount > maxCount) {
            maxCount = currCount;
            modeCount = 1;
        } else if (currCount == maxCount) {
            if (modes != null) {
                modes[modeCount] = currVal;
            }
            modeCount++;
        }
    }

    public int[] solution1(TreeNode root) {
        inorder(root);
        modes = new int[modeCount];
        modeCount = 0;
        currCount = 0;
        inorder(root);
        return modes;
    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        handleValue(root.val);
        inorder(root.right);
    }

    // Solution 2: Morris traversal
    public int[] solution2(TreeNode root) {
        morris(root);
        modes = new int[modeCount];
        modeCount = 0;
        currCount = 0;
        morris(root);
        return modes;
    }

    private void morris(TreeNode root) {
        TreeNode node = root;
        while (node != null) {
            if (node.left == null) {
                handleValue(node.val);
                node = node.right;
            } else {
                TreeNode prev = node.left;
                while (prev.right != null && prev.right != node) {
                    prev = prev.right;
                }
                if (prev.right == null) {
                    prev.right = node;
                    node = node.left;
                } else {
                    prev.right = null;
                    handleValue(node.val);
                    node = node.right;
                }
            }
        }
    }
}
